### ISSUE TITLE
Optimizer.step called after on_epoch_end

### ISSUE THREAD

**Issue by RorryB** [Lightning-AI/pytorch-lightning#21428]
### Bug description

I know this won't effect most people, but i would think on_validation_epoch_end would be getting called after the full training loop has completed, and before the next training loop begins.  But it seems that the ordering is actually:

- training_step
- on_validation_epoch_end
- optimizer_step
- Then into the next training loop

If this is intentional please let me know why this would be the preferred ordering.

### What version are you seeing the problem on?

v2.1

### How to reproduce the bug

```python
import torch
import pytorch_lightning as pl
import os

from torch import nn
from torch.nn import functional as F
from torch.utils.data import TensorDataset, DataLoader, random_split
from torchvision.datasets import MNIST
from torchvision import datasets, transforms

from pytorch_lightning.callbacks import EarlyStopping, ModelCheckpoint

model_path = '.'

class LightningMNISTClassifier(pl.LightningModule):

    def __init__(self):
        super(LightningMNISTClassifier, self).__init__()
        
        # mnist images are (1, 28, 28) (channels, width, height) 
        self.layer_1 = torch.nn.Linear(28 * 28, 4)
        self.layer_2 = torch.nn.Linear(4, 8)
        self.layer_3 = torch.nn.Linear(8, 10)
        self.validation_step_outputs = []
        self.test_step_outputs = []
        
    def forward(self, x):
        batch_size, channels, width, height = x.size()
        
        # (b, 1, 28, 28) -> (b, 1*28*28)
        x = x.view(batch_size, -1)

        # layer 1 (b, 1*28*28) -> (b, 128)
        x = self.layer_1(x)
        x = torch.relu(x)

        # layer 2 (b, 128) -> (b, 256)
        x = self.layer_2(x)
        x = torch.relu(x)

        # layer 3 (b, 256) -> (b, 10)
        x = self.layer_3(x)

        # probability distribution over labels
        x = torch.softmax(x, dim=1)

        return x

    def cross_entropy_loss(self, logits, labels):
        return F.cross_entropy(logits, labels)

    def training_step(self, train_batch, batch_idx):
        print(f"DEBUG: training_step start (batch {batch_idx})", flush=True)
        x, y = train_batch
        logits = self.forward(x)
        loss = self.cross_entropy_loss(logits, y)

        logs = {'train_loss': loss}
        print(f"DEBUG: training_step end (batch {batch_idx})", flush=True)
        return {'loss': loss, 'log': logs}

    def optimizer_step(self, epoch, batch_idx, optimizer, optimizer_closure):
        print(f"DEBUG: optimizer_step called (epoch {epoch}, batch {batch_idx})", flush=True)
        optimizer.step(closure=optimizer_closure)

    def validation_step(self, val_batch, batch_idx):
        x, y = val_batch
        logits = self.forward(x)
        loss = self.cross_entropy_loss(logits, y)
        self.validation_step_outputs.append(loss)
        self.log("val_loss", loss)


    def test_step(self, val_batch, batch_idx):
        x, y = val_batch
        logits = self.forward(x)
        loss = self.cross_entropy_loss(logits, y)
        self.log("test_loss", loss)
        
    def on_validation_epoch_end(self):
        avg_loss = torch.tensor(self.validation_step_outputs).mean()
        validation_step_outputs = []
        tensorboard_logs = {'val_loss': avg_loss}
        print("DEBUG: on_validation_epoch_end finished", flush=True)
        return {'avg_val_loss': avg_loss, 'log': tensorboard_logs}

    def on_test_epoch_end(self):
        avg_loss = torch.tensor(self.test_step_outputs).mean()
        tensorboard_logs = {'test_loss': avg_loss}
        return {'avg_test_loss': avg_loss, 'log': tensorboard_logs}

    def configure_optimizers(self):
        optimizer = torch.optim.Adam(self.parameters(), lr=1e-3)
        lr_scheduler = {'scheduler': torch.optim.lr_scheduler.ExponentialLR(optimizer, gamma = 0.95),
                                        'name': 'expo_lr'}
        return [optimizer], [lr_scheduler]


# Custom Callbacks
class MyPrintingCallback(pl.Callback):

    def on_init_start(self, trainer):
        print('Starting to init trainer!')

    def on_init_end(self, trainer):
        print('trainer is init now')

    def on_train_end(self, trainer, pl_module):
        print('do something when training ends')


# ## Prepare Data

def prepare_data():
        
        
    transform=transforms.Compose([
                    transforms.ToTensor(),
                    transforms.Normalize((0.1307,), (0.3081,))
                    ])
    mnist_train = datasets.MNIST('./data', train=True, download=True,
                                    transform=transform)
    mnist_test = datasets.MNIST('./data', train=False, transform=transform)
    
    mnist_train, mnist_val = random_split(mnist_train, [55000, 5000])

    mnist_test = datasets.MNIST('./data', train=False, download=True, transform=transform)

    return mnist_train, mnist_val, mnist_test


# ### Get train, validation, test data

train, val, test = prepare_data()

# ### Prepare Data Loader

train_loader, val_loader, test_loader = DataLoader(train, batch_size=64), DataLoader(val, batch_size=64), DataLoader(test, batch_size=64)

# ## Train Model

model = LightningMNISTClassifier()

# Set Early Stopping
early_stopping = EarlyStopping('val_loss', mode='min', patience=5)
# saves checkpoints to 'model_path' whenever 'val_loss' has a new min
checkpoint_callback = ModelCheckpoint(dirpath=model_path+'mnist_{epoch}-{val_loss:.2f}',
                                                                            monitor='val_loss', mode='min', save_top_k=3)

trainer = pl.Trainer(max_epochs=30, profiler='simple', 
                                         callbacks=[early_stopping,checkpoint_callback],
                                         default_root_dir=model_path) #gpus=1

trainer.fit(model, train_dataloaders=train_loader, val_dataloaders=val_loader)

# Print model's state_dict
print("Model's state_dict:")
for param_tensor in model.state_dict():
        print(param_tensor, "\t", model.state_dict()[param_tensor].size())

# ## TensorBoard

# # copy lightning logs from google drive to local machine
os.environ['lightning_logs'] = model_path+'lightning_logs'

trainer.test(dataloaders=test_loader, ckpt_path=None)

# ## Inference
PATH = checkpoint_callback.best_model_path
print(PATH)

inference = LightningMNISTClassifier.load_from_checkpoint(PATH)

x = torch.cat((test[0][0],test[1][0],test[2][0]), 0) # 3 image
x = x.unsqueeze(1).to('cuda')

y = [test[0][1],test[1][1],test[2][1]]

import numpy as np

# Do Prediction
logits = inference(x)
print('Prediction :',np.argmax(logits.to('cpu').detach().numpy(), axis=1))
print('Real :', y)
```

### Error messages and logs

DEBUG: training_step start (batch 858)
DEBUG: training_step end (batch 858)
Epoch 0: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████▊| 859/860 [00:02<00:00, 307.02it/s, v_num=0]DEBUG: optimizer_step called (epoch 0, batch 859)
DEBUG: training_step start (batch 859)
DEBUG: training_step end (batch 859)
Epoch 0: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 860/860 [00:02<00:00, 307.16it/s, v_num=0DEBUG: on_validation_epoch_end finished███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 79/79 [00:00<00:00, 399.54it/s]
Epoch 1:   0%|                                                                                                                                                                   | 0/860 [00:00<?, ?it/s, v_num=0]DEBUG: optimizer_step called (epoch 1, batch 0)                                                                                                                                                                   
DEBUG: training_step start (batch 0)
DEBUG: training_step end (batch 0)

### Environment

<details>
  <summary>Current environment</summary>

```
#- PyTorch Lightning Version (e.g., 2.5.0):
#- PyTorch Version (e.g., 2.5):
#- Python version (e.g., 3.12):
#- OS (e.g., Linux):
#- CUDA/cuDNN version:
#- GPU models and configuration:
#- How you installed Lightning(`conda`, `pip`, source):
```

</details>

cc @ethanwharris

**Comment by RorryB** [comment]
Just realized this also happens at epoch 0.  So optimizer.step() is called before any data is processed.

**Comment by GdoongMathew** [comment]
Hi @RorryB , thanks for the report!

From how the hooks are implemented the expected ordering is [`automatic_optimization.run (training_step + optimizer.step)`](https://github.com/Lightning-AI/pytorch-lightning/blob/862314360457fed35672ec9d0c6f5f1998a61eae/src/lightning/pytorch/loops/training_epoch_loop.py#L240) -> [`on_validation_epoch_end`](https://github.com/Lightning-AI/pytorch-lightning/blob/862314360457fed35672ec9d0c6f5f1998a61eae/src/lightning/pytorch/loops/training_epoch_loop.py#L285). Your log output appears to show the same ordering.

Could you share a bit more about whats not working for you or what behavior you expect instead?

**Comment by RorryB** [comment]
Sure, so I would expect the ordering to be

- Automatic_optimization.run
    - training step
    - optimizer step
    - training step
    - optimizer step
    - ......
    - training step
    - optimizer step
- on validation epoch end

but it looks like it is actually 
- Automatic_optimization.run
    - optimizer step
    - training step
    - optimizer step
    - training step
    - ......
    - training step
- on validation epoch end
- Next automatic optimizer.run
    -  optimizer step

i.e. an optimizer step is called before the first training step, which is also repeated on every training epoch.  and then each training epoch ends with a training step that does not have the optimizer.step actually called until the next training epoch.

It might not be coming through clearly from the log because the optimizer.step print is on the same line as the training bar.  But if you just look at the debugs without the training bar the order is:

DEBUG: training_step start (batch 858)
DEBUG: training_step end (batch 858)
DEBUG: optimizer_step called (epoch 0, batch 859)
DEBUG: training_step start (batch 859)
DEBUG: training_step end (batch 859)
on_validation_epoch_end 
DEBUG: optimizer_step called (epoch 1, batch 0)
DEBUG: training_step start (batch 0)
DEBUG: training_step end (batch 0)

**Comment by GdoongMathew** [comment]
Thanks for the clarification. So this is actually an expected behavior caused by `lightning` utilizing the closure mechanism in torch optimizer. A reference could be found in this https://github.com/Lightning-AI/pytorch-lightning/pull/21120#issuecomment-3227587883 as well.

Basically, `lightning` wraps the work from `training_step` through `on_before_optimizer_step` into one big closure function and passes that to `optimizer.step` argument. Optimizer would call that closure internally ([ref](https://github.com/pytorch/pytorch/blob/main/torch/optim/sgd.py#L114-L116)). That's why the logs can look like `optimizer.step` runs before `training_step`, but the training logic inside the closure is still executed before the actual parameter update.

**Comment by RorryB** [comment]
Im not sure if I follow.  From your referenced comment:

# closure function for optimizer
            strategy.training_step()
            └── LightningModule.training_step()

            callback.on_before_zero_grad()
            LightningModule.on_before_zero_grad()
            LightningModule.optimizer_zero_grad()
            strategy.backward()
            ├── callback.on_before_backward()
            ├── LightningModule.on_before_backward()
            ├── LightningModule.backward()
            │   └── loss.backward()
            ├── callback.on_after_backward()
            └── LightningModule.on_after_backward()

            callback.on_before_optimizer_step()
            LightningModule.on_before_optimizer_step()
            LightningModule.configure_gradient_clipping()
    
            optimizer.step(closure)

It still looks like that (training_step through on_before_optimizer_step) is fully called before optimizer.step.  I don't understand why the logs would make it look like optimizer.step() is running first if it is not.  My actual use case where i found this is also more complex than just printing the logs and it was causing things to crash. I can try to add complexity to this minimum example to cause the same effect with more than just print statements.

**Comment by GdoongMathew** [comment]
So, at a high level, the pseudo-code for the optimizer closure looks like this:

```python
def closure():
    LightningModule.training_step()
    LightningModule.optimizer_zero_grad()
    loss.backward()
    LightningModule.on_before_optimizer_step()

optimizer.step(closure=closure)
```

If you add a debug print before and after the `optimizer.step`

```python
class LightningMNISTClassifier(pl.LightningModule):
    def optimizer_step(self, epoch, batch_idx, optimizer, optimizer_closure):
        print(f"DEBUG: optimizer_step called (epoch {epoch}, batch {batch_idx})", flush=True)
        optimizer.step(closure=optimizer_closure)    # --> The `training_step` is included in the `optimizer_closure` here
        print(f"DEBUG: optimizer_step called finished (epoch {epoch}, batch {batch_idx})", flush=True)
```

Then you should see logs similar to the following:
```text
DEBUG: optimizer_step called (epoch 0, batch 858)
DEBUG: training_step start (batch 858)
DEBUG: training_step end (batch 858)
DEBUG: optimizer_step called finished (epoch 0, batch 858)
```
 which shows the training step is executed within the optimizer step via the provided closure, rather than being called separately afterward.

**Comment by RorryB** [comment]
Awesome, thank you for this detailed explanation!  This helped me track down my bug.  I've never seen this step(closure) idea before.  

FYI, and If anyone else looks at this later with the same problem, I have fairly complicated setup for my project that does some other stuff before the step().  My usual order without Lightning is training step -> zero grad -> other stuff -> optimizer step.  But closure was re ordering it to be other stuff -> training step -> zero grad -> optimizer step.  By taking this closure concept into account I was able to rearrange where my other stuff is running and get things in the right order.

Thanks again!
